modules/xmpp/JingleSessionPC.js

```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
+        let removeSsrcInfo = this._parseSsrcInfoFromSourceRemove(elem, sdp);
+
+        let newRemoteSdp = this._processRemoteRemoveSource(removeSsrcInfo);
+        this._renegotiate(newRemoteSdp)
+            .then(() => {
+                logger.info("Remote source-remove processed");
+                var newSdp = new SDP(this.peerconnection.localDescription.sdp);
+                logger.log("SDPs", mySdp, newSdp);
+                this.notifyMySSRCUpdate(mySdp, newSdp);
+                finishedCallback();
+            }, (error) => {
+                logger.info("Error renegotiating after processing remote source-remove: " + error);
+                finishedCallback(error);
+            });
+    };
+    this.modificationQueue.push(workFunction);
+};
+
+/**
+ * The 'task' function will be given a callback it MUST call with either:
+ *  1) No arguments if it was successful or
+ *  2) An error argument if there was an error
+ * If the task wants to process the success or failure of the task, it should
+ *  pass a handler to the .push function, e.g.:
+ * queue.push(task, (err) => {
+ *     if (err) {
+ *         // error handling
+ *     } else {
+ *         // success handling
+ *     }
+ * });
+ */
+JingleSessionPC.prototype._processQueueTasks = function (task, finishedCallback) {
+    task(finishedCallback);
+};
+
+/**
+ * Takes in a jingle offer iq, returns the new sdp offer
+ */
+JingleSessionPC.prototype._processNewJingleOfferIq = function(offerIq) {
+    let remoteSdp = new SDP('');
+    if (this.webrtcIceTcpDisable) {
+        remoteSdp.removeTcpCandidates = true;
+    }
+    if (this.webrtcIceUdpDisable) {
+        remoteSdp.removeUdpCandidates = true;
+    }
+    if (this.failICE) {
+        remoteSdp.failICE = true;
+    }
+
+    remoteSdp.fromJingle(offerIq);
+    this.readSsrcInfo($(offerIq).find(">content"));
+    return remoteSdp;
+};
+
+JingleSessionPC.prototype._processRemoteRemoveSource = function (removeSsrcInfo) {
+    let remoteSdp = new SDP(this.peerconnection.remoteDescription.sdp);
+    removeSsrcInfo.forEach(function(lines, idx) {
+        lines = lines.split('\r\n');
+        lines.pop(); // remove empty last element;
+        lines.forEach(function(line) {
+            remoteSdp.media[idx] = remoteSdp.media[idx].replace(line + '\r\n', '');
+        });
+    });
+    remoteSdp.raw = remoteSdp.session + remoteSdp.media.join('');
+
+    return remoteSdp;
+};
 
-    $(elem).each(function (idx, content) {
+JingleSessionPC.prototype._processRemoteAddSource = function (addSsrcInfo) {
+    let remoteSdp = new SDP(this.peerconnection.remoteDescription.sdp);
+    addSsrcInfo.forEach(function(lines, idx) {
+        remoteSdp.media[idx] += lines;
+    });
+    remoteSdp.raw = remoteSdp.session + remoteSdp.media.join('');
+
+    return remoteSdp;
+};
+
+/**
+ * Do a new o/a flow using the existing remote description
+ * @param sdp: optional remote sdp to use.  if not provided, the remote
```

Comment:
It's good to have the exact name of the parameter and type:
`@param {Type} exampleVar <description>`

Also `@returns` is missing.